src/addons/ReactFragment.js

```diff
@@ -121,71 +63,35 @@ var ReactFragment = {
       }
-      if (canWarnForReactFragment) {
-        var proxy = {};
-        Object.defineProperty(proxy, fragmentKey, {
-          enumerable: false,
-          value: object,
-        });
-        Object.defineProperty(proxy, didWarnKey, {
-          writable: true,
-          enumerable: false,
-          value: false,
-        });
-        for (var key in object) {
-          proxyPropertyAccessWithWarning(proxy, key);
-        }
-        Object.preventExtensions(proxy);
-        return proxy;
-      }
     }
-    return object;
-  },
-  // Extract the original keyed object from the fragment opaque type. Warn if
-  // a plain object is passed here.
-  extract: function(fragment) {
-    if (__DEV__) {
-      if (canWarnForReactFragment) {
-        if (!fragment[fragmentKey]) {
-          var fragmentKeys = getFragmentKeyString(fragment);
+    invariant(
+      object.nodeType !== 1,
+      'React.addons.createFragment(...): Encountered an invalid child; DOM ' +
+      'elements are not valid children of React components.'
+    );
+
+    var result = [];
+    var context = {
+      result: result,
+      prefix: '',
+    };
+
+    for (var key in object) {
+      if (__DEV__) {
+        if (!warnedAboutNumeric && numericPropertyRegex.test(key)) {
           warning(
-            didWarnForFragment(fragmentKeys),
-            'Any use of a keyed object should be wrapped in ' +
-            'React.addons.createFragment(object) before being passed as a ' +
-            'child. {%s}',
-            fragmentKeys
+            false,
+            'React.addons.createFragment(...): Child objects should have ' +
+            'non-numeric keys so ordering is preserved.'
           );
-          return fragment;
-        }
-        return fragment[fragmentKey];
-      }
-    }
-    return fragment;
-  },
-  // Check if this is a fragment and if so, extract the keyed object. If it
-  // is a fragment-like object, warn that it should be wrapped. Ignore if we
-  // can't determine what kind of object this is.
-  extractIfFragment: function(fragment) {
-    if (__DEV__) {
-      if (canWarnForReactFragment) {
-        // If it is the opaque type, return the keyed object.
-        if (fragment[fragmentKey]) {
-          return fragment[fragmentKey];
-        }
-        // Otherwise, check each property if it has an element, if it does
-        // it is probably meant as a fragment, so we can warn early. Defer,
-        // the warning to extract.
-        for (var key in fragment) {
-          if (fragment.hasOwnProperty(key) &&
-              ReactElement.isValidElement(fragment[key])) {
-            // This looks like a fragment object, we should provide an
-            // early warning.
-            return ReactFragment.extract(fragment);
-          }
+          warnedAboutNumeric = true;
         }
       }
+      context.prefix = escapeUserProvidedKey(key) + '/';
+      traverseAllChildren(object[key], processSingleChildWithContext, context);
```

Comment:
I don't think that this is quite equivalent for nested fragments, is it?

```js
var foo = React.createFragment({ foo: <Foo /> });
var bar1 = React.createFragment({
  hi: <div />,
  bar: foo
});
var bar2 = React.createFragment({
  bar: foo,
  bye: <span />
});
```

Does swapping between `bar1` and `bar2` preserve state of `<Foo />`?

Maybe we should copy/paste a custom `traverseAllChildren` here to preserve legacy behavior? (and also removing the dependency on React internals)